runtime/src/comm/ugni/comm-ugni.c

```diff
@@ -4025,19 +4027,31 @@ void chpl_comm_pre_task_exit(int all)
   if (all) {
     if (chpl_nodeID == 0) {
       for (int i = 1; i < chpl_numNodes; i++) {
-        fork_shutdown(i);
+        fork_shutdown(i, true /*coordinated_shutdown*/);
       }
     } else {
       chpl_wait_for_shutdown();
     }
 
-    chpl_comm_barrier("chpl_comm_pre_task_exit");
+    if (coordinated_shutdown) {
+      chpl_comm_barrier("chpl_comm_pre_task_exit");
 
-    polling_task_please_exit = true;
+      polling_task_please_exit = true;
 
-    if (chpl_nodeID == 0) {
-      while (!polling_task_done)
-        sched_yield();
+      if (chpl_nodeID == 0) {
+        while (!polling_task_done)
+          sched_yield();
+      }
+    }
+  } else {
```

Comment:
I don't understand this logic. Under what circumstances is `all` `false`, and why is an uncoordinated shutdown done in that case?

Reply:
`all` means "are all locales participating in this shutdown, or just one" `all=false` when calling `chpl_exit_any`, which is called for `exit()` and `halt()`. `all=false` is by definition an uncoordinated shutdown because one locale is trying to shut the entire program down.

`all=true` happens for normal shutdown when a chapel program is exiting. The control flow for how this all works is quite confusing, but I can walk you through it if that'd be helpful.